### PR TITLE
refactor(excel2json): new excel reader (DEV-3049)

### PR DESCRIPTION
**Pull request by Nora-Olivia-Ammann**

### CODE REVIEW

File: src/dsp_tools/commands/excel2json/resources.py [hunk 1]
```diff
@@ -197,47 +191,50 @@ def excel2resources(
             and the success status (True if everything went well)
     """
 
-    # load file
-    all_classes_df = read_and_clean_excel_file(excelfile=excelfile)
-    all_classes_df = prepare_dataframe(
-        df=all_classes_df,
+    df_dict = read_and_clean_all_sheets_excelfile(excelfile)
+    classes_df = df_dict["classes"]
+    df_dict.pop("classes")
```
Comment (BalduinLandolt) [inline src/dsp_tools/commands/excel2json/resources.py:196]:
```suggestion
    classes_df = df_dict.pop("classes")
```

I may be wrong, but I think that should work?

File: src/dsp_tools/commands/excel2json/resources.py [hunk 2]
```diff
@@ -197,47 +191,50 @@ def excel2resources(
...
+        )
+    duplicate_check = check_column_for_duplicate(classes_df, "name")
+    if duplicate_check:
+        msg = "The excel file 'resources.xlsx', sheet 'classes' has a problem.\n"
+        msg += duplicate_check.execute_error_protocol()
+        raise InputError(msg)
+    # check that all the sheets have an entry in the names column and vice versa
+    if (all_names := set(classes_df["name"].tolist())) != (all_sheets := set(df_dict.keys())):
+        msg = ResourcesExcelSheetsNotAsExpected(all_names, all_sheets).execute_error_protocol()
+        raise InputError(msg)
```
Comment (BalduinLandolt) [inline src/dsp_tools/commands/excel2json/resources.py:240]: how would you feel about returning the Problem and then raising the error on the caller side?

File: src/dsp_tools/commands/excel2json/utils.py [hunk 3]
```diff
@@ -39,6 +44,39 @@ def read_and_clean_excel_file(excelfile: str, sheetname: str | int = 0) -> pd.Da
     return read_df
 
 
+def read_and_clean_all_sheets_excelfile(excelfile: str) -> dict[str, pd.DataFrame]:
```
Comment (BalduinLandolt) [inline src/dsp_tools/commands/excel2json/utils.py:47]: how about renaming this to `read_and_clean_all_sheets()` without the `excelfile` which seems a bit redundant

File: src/dsp_tools/commands/excel2json/utils.py [hunk 4]
```diff
@@ -39,6 +44,39 @@ def read_and_clean_excel_file(excelfile: str, sheetname: str | int = 0) -> pd.Da
...
+        # (unclear which formatting properties exactly).
+        # Apparently, the excel2json test files have one of the unsupported formatting properties.
+        # Credits: https://stackoverflow.com/a/70537454/14414188
+        with mock.patch("openpyxl.styles.fonts.Font.family.max", new=100):
+            df_dict = pd.read_excel(excelfile, sheet_name=None)
+    try:
+        return {name.strip(""): clean_data_frame(df) for name, df in df_dict.items()}
+    except AttributeError:
+        msg = InvalidExcelSheetNameProblem(excelfile, list(df_dict.keys())).execute_error_protocol()
+        raise InputError(msg) from None
```
Comment (BalduinLandolt) [inline src/dsp_tools/commands/excel2json/utils.py:77]: as above

**Review by BalduinLandolt — APPROVED** [review]
some minor remarks, otherwise looking good

File: src/dsp_tools/commands/excel2json/resources.py [hunk 5]
```diff
@@ -197,47 +191,50 @@ def excel2resources(
...
 
     if path_to_output_file:
         with open(file=path_to_output_file, mode="w", encoding="utf-8") as file:
             json.dump(resources, file, indent=4, ensure_ascii=False)
             print(f"resources section was created successfully and written to file '{path_to_output_file}'")
 
     return resources, True
+
+
+def _validate_excel_file(classes_df: pd.DataFrame, df_dict: dict[str, pd.DataFrame]) -> None:
```
Comment (jnussbaum) [inline src/dsp_tools/commands/excel2json/resources.py:220]: `df_dict` doesn't tell me what is inside. `resource_dfs` would be more telling

File: src/dsp_tools/commands/excel2json/utils.py [hunk 6]
```diff
@@ -39,6 +44,39 @@ def read_and_clean_excel_file(excelfile: str, sheetname: str | int = 0) -> pd.Da
     return read_df
 
 
+def read_and_clean_all_sheets_excelfile(excelfile: str) -> dict[str, pd.DataFrame]:
+    """
+    This function reads an Excel file with all its sheets,
+    if there is a ValueError then it patches the openpyxl part that creates the
+    error and opens it with that patch.
+    It cleans and then returns a dictionary with the dataframes.
```
Comment (jnussbaum) [inline src/dsp_tools/commands/excel2json/utils.py:52]:
```suggestion
    This function reads an Excel file with all its sheets.
    If there is a ValueError, it patches the openpyxl part that causes the error
    and opens it with that patch.
    It cleans the dataframes and then returns them in the form {sheet_name: dataframe}.
```

File: src/dsp_tools/commands/excel2json/utils.py (hunk 3 above)
Comment (jnussbaum) [inline src/dsp_tools/commands/excel2json/utils.py:47]: The `excelfile` in the function name is redundant.

File: src/dsp_tools/commands/excel2json/utils.py [hunk 7]
```diff
@@ -39,6 +44,39 @@ def read_and_clean_excel_file(excelfile: str, sheetname: str | int = 0) -> pd.Da
...
+
+    Args:
+        excelfile: The name of the Excel file
```
Comment (jnussbaum) [inline src/dsp_tools/commands/excel2json/utils.py:55]: That's trivial. A more informative docstring would be:

```suggestion
        excelfile: path to the Excel file
```

File: src/dsp_tools/commands/excel2json/utils.py [hunk 8]
```diff
@@ -39,6 +44,39 @@ def read_and_clean_excel_file(excelfile: str, sheetname: str | int = 0) -> pd.Da
...
+
+    Returns:
+        A dictionary with all the sheets as dataframes
```
Comment (jnussbaum) [inline src/dsp_tools/commands/excel2json/utils.py:58]: Could you tell that more precise? e.g. 

```suggestion
        All sheets of the excel file, in the form of a dictionary {sheet_name: dataframe}
```

**Review by jnussbaum — APPROVED** [review]
LGTM

File: src/dsp_tools/commands/excel2json/input_error.py [hunk 9]
```diff
@@ -114,6 +114,60 @@ def execute_error_protocol(self) -> str:
         )
 
 
+@dataclass(frozen=True)
+class InvalidExcelSheetNameProblem:
+    """This class contains information if the excel sheet names are not strings."""
```
[the remark this answers is not in the document]
Comment (Nora-Olivia-Ammann) [reply src/dsp_tools/commands/excel2json/input_error.py:119]: I prefer to be on the safe side. Rather than having a conversion crashed.

**Pull request merged**